src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.mesos.scheduler.plan.Block;
+import org.apache.mesos.scheduler.repair.constrain.LaunchConstrainer;
+import org.apache.mesos.scheduler.repair.monitor.FailureMonitor;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides the core functionality of the automatic repair mechanisms.
+ */
+public class RepairScheduler {
+    private final Log log = LogFactory.getLog(getClass());
+
+    private final String targetConfigName;
+    private final CoreFrameworkState frameworkState;
+    private final OfferAccepter offerAccepter;
+    private final RepairOfferRequirementProvider offerReqProvider;
+    private final FailureMonitor failureMonitor;
+    private final LaunchConstrainer launchConstrainer;
+    private final AtomicReference<RepairStatus> repairStatusRef;
+
+    public RepairScheduler(
+            String targetConfigName,
+            CoreFrameworkState frameworkState,
+            RepairOfferRequirementProvider offerReqProvider,
+            OfferAccepter offerAccepter,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor,
+            AtomicReference<RepairStatus> repairStatusRef) {
+        this.targetConfigName = targetConfigName;
+        this.frameworkState = frameworkState;
+        this.offerReqProvider = offerReqProvider;
+        this.offerAccepter = offerAccepter;
+        this.failureMonitor = failureMonitor;
+        this.launchConstrainer = launchConstrainer;
+        this.repairStatusRef = repairStatusRef;
+    }
+
+    public List<OfferID> resourceOffers(SchedulerDriver driver, List<Offer> offers, Block block)
+            throws Exception {
+        List<OfferID> acceptedOffers = new ArrayList<OfferID>();
+        List<TaskInfo> terminatedTasks = getTerminatedTasks(block);
+
+        // Compute the stopped and failed pool for UI
+        RepairStatus oldStatus = repairStatusRef.get();
+        if (oldStatus == null) {
+            oldStatus = new RepairStatus(Collections.emptyList(), Collections.emptyList());
+        }
+        List<String> oldStopped = oldStatus.getStopped();
+        List<String> newStopped = new ArrayList<>(terminatedTasks.stream()
+                .filter(failureMonitor::hasFailed)
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList()));
+        newStopped.addAll(oldStopped);
+        //TODO (person) possibly write this list down in ZK, to ensure the UI doesn't change upon restart
+        newStopped = newStopped.stream().distinct().collect(Collectors.toList());
+        List<String> newFailed = terminatedTasks.stream()
+                .filter(it -> !failureMonitor.hasFailed(it))
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList());
+
+        OfferRequirement offerReq = null;
+        boolean replacingMissing = false;
+
+        if (terminatedTasks.size() > 0) {
```

Comment:
One thing we discussed at the project kickoff was the challenges of composing scheduler logic together. That is one of the requirements that set us down the txnplan path. Once you decided that the simpler approach was necessary, we reverted back to this design, as we had discussed. I brought up my concern with this when I mentioned the need to thoughtfully created scheduler combinators, to express different sorts of logic.

If you have an alternative design you'd prefer for this logic, if it's simple I should be able to do it by the end of the week; otherwise, I can come up with a timeline for you.